### ISSUE TITLE
`BeginBlock` and `EndBlock` events handling

### ISSUE THREAD

**Issue by RiccardoM** [confio/ts-relayer#188]
### Context
Currently the relayer is able to relay only the packets that can be found by searching for transactions events. However there are some chains (ie. [Band Protocol](https://github.com/bandprotocol/chain)) that rely on the `BeginBlock` and `EndBlock` methods to emit events that would not be otherwise emitted. 

This is the case of Band's `x/oracle` module that uses the `EndBlock` method to aggregate all the results coming from different reporters, execute the on-chain oracle script and then store the result. Once the result is stored, if the request was performed though an IBC connection the `EndBlock` sends an IBC packet to the source chain to make sure they get the result properly without having to perform a new query later on. 

This method however is not supported by the TS Relayer since it uses the following method to get the events: 
```typescript
public async querySentPackets({minHeight, maxHeight}: QueryOpts = {}): Promise<PacketWithMetadata[]> {
    let query = `send_packet.packet_connection='${this.connectionID}'`;
    if (minHeight) {
      query = `${query} AND tx.height>=${minHeight}`;
    }
    if (maxHeight) {
      query = `${query} AND tx.height<=${maxHeight}`;
    }

    const search = await this.client.tm.txSearchAll({ query });
    
    ...
}
```

### Proposed solution
A possible solution could be the one of using the `/block_results?height={height}` endpoint, querying all the blocks that were not parsed previously to get the `begin_block_events` and `end_block_events` properly. This however would mean basing the solution on a high number of queries of an unreliable part of Tendermint. 

### CCs
CCing some people that might be interested. Particularly: 
- @ethanfrey and @orkunkl
- @taobun and @Benzbeeb (Band developers)

**Comment by RiccardoM** [comment]
Speaking with the Tendermint team (particularly @marbar who I thank) I found out that the following RPC endpoint has been introduced in Tendermint v0.34.x: 

```
/block_search?query=
```

This has been introduced with [PR #6626](https://github.com/tendermint/tendermint/pull/6226), made available with [`v0.34.9`](https://github.com/tendermint/tendermint/blob/master/CHANGELOG.md#features) and is already available inside the Cosmos SDK with [`v0.42.4`](https://github.com/cosmos/cosmos-sdk/blob/master/CHANGELOG.md#improvements-1)

**Comment by ethanfrey** [comment]
Thank you for raising this issue.

I remember when the question from band first came up on discord and I motivated the Tendermint team to add the begin/end block events to the event indexer (before they were only available via subscription not search). I was not sure it made it into the SDK, nor did I remember to follow up. 

I agree we should add support. My only concern is that such queries will produce errors on tendermint < v0.34.9, meaning it may cause the relayer to fail to support eg. v0.41.1 chains. Is band on v0.42.4+?

Should we keep backwards compatibility for v0.41.1+ chains (as now) or just require v0.42.4 as a minimum requirement (are there any mainnets that are running v0.41??)

**Comment by RiccardoM** [comment]
> I agree we should add support. My only concern is that such queries will produce errors on tendermint < v0.34.9, meaning it may cause the relayer to fail to support eg. v0.41.1 chains. Is band on v0.42.4+?

Yes, Band is v0.42.4 based. And all Cosmos chains running on v0.42.x should be as well since it implements some security bug fix that were implemented with Tendermint v0.34.9 (the same version that introduces this endpoint).

> Should we keep backwards compatibility for v0.41.1+ chains (as now) or just require v0.42.4 as a minimum requirement (are there any mainnets that are running v0.41??)

To be honest, I have no idea if there are chains running on v0.41.1. As as as I remember, the upgrade to v0.42 can be made without any particular migration needed. I think all chains should be based on v0.42.4

**Comment by ethanfrey** [comment]
I guess this means we can just update the documentation (in your PR that breaks support for older version). Which is easier than handing fallback for older versions
